src/main/java/jenkins/scm/api/SCMSource.java

```diff
@@ -155,21 +181,104 @@ protected final SCMSourceCriteria getCriteria() {
+                                                        @CheckForNull TaskListener listener)
+            throws IOException, InterruptedException {
+        _retrieve(criteria, observer, defaultListener(listener));
+        return observer;
+    }
+
+    /**
+     * Fetches the latest heads and corresponding revisions. Implementers are free to cache intermediary results
+     * but the call must always check the validity of any intermediary caches.
+     *
+     * @param observer an optional observer of interim results.
+     * @param listener the task listener.
+     * @throws IOException if an error occurs while performing the operation.
+     * @throws InterruptedException if any thread has interrupted the current thread.
+     */
+    protected void retrieve(@NonNull SCMHeadObserver observer, @NonNull TaskListener listener)
+            throws IOException, InterruptedException {
+        if (MethodUtils.isOverridden(SCMSource.class, getClass(), "retrieve",
+                SCMSourceCriteria.class, SCMHeadObserver.class, TaskListener.class)) {
+            retrieve(getCriteria(), observer, listener);
+        } else {
+            throw new AbstractMethodError("Implement retrieve(SCMSourceCriteria,SCMHeadObserver,TaskListener)");
+        }
+    }
+
+
+    /**
+     * Fetches the latest heads and corresponding revisions. Implementers are free to cache intermediary results
+     * but the call must always check the validity of any intermediary caches.
+     *
+     * @param observer an optional observer of interim results.
+     * @param listener the task listener.
+     * @throws IOException if an error occurs while performing the operation.
+     * @throws InterruptedException if any thread has interrupted the current thread.
+     */
+    private void _retrieve(@CheckForNull SCMSourceCriteria criteria,
+                           @NonNull SCMHeadObserver observer,
+                           @NonNull TaskListener listener)
+            throws IOException, InterruptedException {
+        if (MethodUtils.isOverridden(SCMSource.class, getClass(), "retrieve",
+                SCMSourceCriteria.class, SCMHeadObserver.class, TaskListener.class)) {
+            // w00t this is a new implementation
+            retrieve(criteria, observer, listener);
+        } else if (MethodUtils.isOverridden(SCMSource.class, getClass(), "retrieve",
+                SCMHeadObserver.class, TaskListener.class)){
+            // oh dear, legacy implementation
+            SCMSourceCriteria hopefullyNull = compatibilityHack.get();
+            compatibilityHack.set(criteria == null ? nullSCMSourceCriteria : criteria);
+            try {
+                retrieve(observer, listener);
+            } finally {
+                if (hopefullyNull != null) {
+                    // performance is going to be painful if you are nesting them
+                    compatibilityHack.set(hopefullyNull);
+                } else {
+                    compatibilityHack.remove();
```

Comment:
Not quite. `set(null)` will leave an entry in the `ThreadLocalMap` whereas `remove()` clears the entry.